splito/lohi/_lo.py

```diff
@@ -0,0 +1,169 @@
+from rdkit import DataStructs
+import numpy as np
+from tqdm import tqdm
+import datamol as dm
+import functools
+from loguru import logger
+
+
+class LoSplitter:
+    def __init__(
+        self,
+        threshold: float = 0.4,
+        min_cluster_size: int = 5,
+        max_clusters: int = 50,
+        std_threshold: float = 0.60,
+    ):
+        """
+        A splitter that prepares data for training ML models for Lead Optimization or to guide
+        molecular generative models. These models must be sensitive to minor modifications of
+        molecules, and this splitter constructs a test that allows the evaluation of a model's
+        ability to distinguish those modifications.
+
+        Args:
+            threshold: ECFP4 1024-bit Tanimoto similarity threshold.
+                Molecules more similar than this threshold are considered too similar and can be grouped together in one cluster.
+            min_cluster_size: the minimum number of molecules per cluster.
+            max_clusters: the maximum number of selected clusters. The remaining molecules go to the training set.
+                This can be useful for limiting your test set to get more molecules in the train set.
+            std_threshold: the lower bound of the acceptable standard deviation for a cluster's values. It should be greater than the measurement noise.
+                For ChEMBL-like data set it to 0.60 for logKi and 0.70 for logIC50.
+                Set it lower if you have a high-quality dataset.
+
+        For more information, see a tutorial in the docs and Steshin 2023, Lo-Hi: Practical ML Drug Discovery Benchmark.
+        """
+        self.threshold = threshold
+        self.min_cluster_size = min_cluster_size
+        self.max_clusters = max_clusters
+        self.std_threshold = std_threshold
+
+    def split(
+        self, smiles: list[str], values: list[float], n_jobs: int = -1, verbose: int = 1
+    ) -> tuple[list[int], list[list[int]]]:
+        """
+        Split the dataset into test clusters and train.
+
+        Args:
+            smiles: list of smiles.
+            values: list of their continuous activity values.
+            verbose: set to 0 to turn off progressbar.
+
+        Returns:
+            train_idx: list of train indices.
+            clusters_idx: list of lists of cluster indices.
+        """
+        if not isinstance(smiles, np.ndarray):
+            smiles = np.array(smiles)
+        if not isinstance(values, np.ndarray):
+            values = np.array(values)
+
+        train_idx, clusters_idx, central_nodes = self._select_distinct_clusters(
+            smiles, values, n_jobs, verbose
+        )
+        train_idx = list(train_idx) + central_nodes
+
+        if not clusters_idx:
+            logger.warninig("No clusters were found. Was your std_threshold too constrained?")
+
+        return train_idx, clusters_idx
```

Comment:
Why do you return the `cluster_idx` as a list (i.e. you essentially return multiple test sets)? I can think of two suggestions here: 

1. If you're supposed to evaluate each cluster independently, maybe we can do: 
```suggestion
        for cluster in cluster_idx: 
            yield train_idx, clusters_idx
```

2. If this is not the case, could we just return all indices as if it is a single test case? 

```suggestion
        return train_idx, np.flatten(clusters_idx)
```

Maybe we could make this configurable through a class parameter?